server/src/main/java/org/opensearch/index/store/remote/filecache/FileCacheCleaner.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.store.remote.filecache;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.apache.logging.log4j.util.Supplier;
+import org.opensearch.common.settings.Settings;
+import org.opensearch.common.xcontent.NamedXContentRegistry;
+import org.opensearch.env.NodeEnvironment;
+import org.opensearch.index.IndexModule;
+import org.opensearch.index.shard.IndexEventListener;
+import org.opensearch.index.shard.ShardId;
+import org.opensearch.index.shard.ShardStateMetadata;
+
+import java.io.IOException;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+
+import static org.opensearch.index.store.remote.directory.RemoteSnapshotDirectoryFactory.LOCAL_STORE_LOCATION;
+
+/**
+ * IndexEventListener to clean up file cache when the index is deleted
+ *
+ * @opensearch.internal
+ */
+public class FileCacheCleaner implements IndexEventListener {
+    private static final Logger log = LogManager.getLogger(FileCacheCleaner.class);
+
+    private final NodeEnvironment nodeEnvironment;
+    private final FileCache fileCache;
+
+    public FileCacheCleaner(NodeEnvironment nodeEnvironment, FileCache fileCache) {
+        this.nodeEnvironment = nodeEnvironment;
+        this.fileCache = fileCache;
+    }
+
+    private Path getShardPath(ShardId shardId) throws IOException {
+        final Path[] paths = nodeEnvironment.availableShardPaths(shardId);
```

Comment:
We can potentially use the change from https://github.com/opensearch-project/OpenSearch/pull/6350/files#diff-64eed71f060448867b2456f621ebd72a43d52cd20db9080a82308978a0930a5a.

Reply:
Moving this to another PR as per Andrew's comment below, will address in there.